packages/liveblocks-react-ui/src/slate/plugins/auto-links.ts

```diff
@@ -137,6 +139,38 @@ function endsWithPeriodOrQuestionMark(textContent: string): boolean {
   );
 }
 
+/**
+ * Helper function to get the "logical length" of a URL, taking into account things like opening/closing parentheses
+ * @param url The URL to check
+ * @returns The "logical length" of the URL
```

Comment:
Out of curiosity, what kind of URLs are being analysed here?

Reply:
Any URL being composed that matched `URL_REGEX` above, so we already know they are valid URLs when this helper is invoked.

Reply:
I was specifically curious about the need for doing the balanced parentheses counting. What's that needed for?

Reply:
It allows you to type "Hello ([https://google.com/hello(world)](https://google.com/hello(world)))". Otherwise, it would either be:
- parentheses are not handled in links
- once you start typing a link, it can't be followed by a closing parenthesis because it would still be counted as part of the link

Reply:
Ohh, that's quite nice, and seems to be what GitHub is also doing! 💯